src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java

```diff
@@ -0,0 +1,258 @@
+  }
+
+  private ModuleFileGlobals execModuleFile(byte[] moduleFile, ModuleKey moduleKey,
+      StarlarkSemantics starlarkSemantics, Environment env)
+      throws ModuleFileFunctionException, InterruptedException {
+    StarlarkFile starlarkFile =
+        StarlarkFile.parse(ParserInput.fromUTF8(moduleFile, moduleKey + "/MODULE.bazel"));
+    if (!starlarkFile.ok()) {
+      Event.replayEventsOn(env.getListener(), starlarkFile.errors());
+      throw errorf("error parsing MODULE.bazel file for %s", moduleKey);
+    }
+
+    ModuleFileGlobals moduleFileGlobals = new ModuleFileGlobals();
+    try (Mutability mu = Mutability.create("module file", moduleKey)) {
+      net.starlark.java.eval.Module predeclaredEnv = getPredeclaredEnv(moduleFileGlobals,
+          starlarkSemantics);
+      Program program = Program.compileFile(starlarkFile, predeclaredEnv);
+      // TODO: check that `program` has no `def`, `if`, etc
+      StarlarkThread thread = new StarlarkThread(mu, starlarkSemantics);
+      thread.setPrintHandler(Event.makeDebugPrintHandler(env.getListener()));
+      Starlark.execFileProgram(program, predeclaredEnv, thread);
+    } catch (SyntaxError.Exception | EvalException e) {
+      throw new ModuleFileFunctionException(e);
+    }
+    return moduleFileGlobals;
+  }
+
+  private static class GetModuleFileResult {
+
+    byte[] moduleFileContents;
+    // `registry` can be null if this module has a non-registry override.
+    @Nullable Registry registry;
+  }
+
+  @Nullable
+  private Optional<GetModuleFileResult> getModuleFile(ModuleKey key, StarlarkOverrideApi override,
+      Environment env) throws ModuleFileFunctionException, InterruptedException {
+    if (override instanceof NonRegistryOverride) {
+      // TODO: The repo name is not necessarily the same as the module name here. Fix later.
+      String repoName = key.getName();
+      RepositoryDirectoryValue repoDir = (RepositoryDirectoryValue) env.getValue(
+          RepositoryDirectoryValue.key(RepositoryName.createFromValidStrippedName(repoName)));
+      if (repoDir == null) {
+        return null;
+      }
+      RootedPath moduleFilePath = RootedPath.toRootedPath(
+          Root.fromPath(repoDir.getPath()), PathFragment.create("MODULE.bazel"));
+      if (env.getValue(FileValue.key(moduleFilePath)) == null) {
+        return Optional.empty();
+      }
+      GetModuleFileResult result = new GetModuleFileResult();
+      result.moduleFileContents = readFile(moduleFilePath.asPath());
+      return Optional.of(result);
+    }
+
+    List<String> registries = Objects.requireNonNull(REGISTRIES.get(env));
+    if (override instanceof RegistryOverride) {
+      String overrideRegistry = ((RegistryOverride) override).getRegistry();
+      if (!overrideRegistry.isEmpty()) {
+        registries = ImmutableList.of(overrideRegistry);
+      }
+    } else if (override != null) {
+      throw errorf("unrecognized override type %s", override.getClass().getName());
```

Comment:
Shouldn't this be an IllegalStateException? (AFAIU the user of Bazel can't do anything to elicit this error, can they?)

(and if they can, dumping raw class names on stderr is probably a bad idea because those are internal to Bazel)